File: 219_ContainsDuplicateII/219_ContainsDuplicateII.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {boolean}
 */

var containsNearbyDuplicate = function (nums, k) {
  let cache = {};
  for (let i = 0; i < nums.length; i++) {
    if (cache[nums[i]] != undefined && i - cache[nums[i]] <= k) {
      return true;
    }
    cache[nums[i]] = i;
  }
  return false;
};

var containsNearbyDuplicate = function (nums, k) {
  let map = new Map();
  for (let i = 0; i < nums.length; i++) {
    if (map.has(nums[i]) && i - map.get(nums[i]) <= k) {
      return true;
    }
    map.set(nums[i], i);
  }
  return false;
};

//map is better computationally for larger data sets
//functions in log n and no hash collisioons

File: 219_ContainsDuplicateII/219_ContainsDuplicateII.py
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        cache = dict()
        for i, n in enumerate(nums):
            if n in cache and abs(i-cache[n]) <= k:
                return True
            else: 
                cache[n] = i
        return False
